Next major version

line-o opened the pull request:
The feedback from the community 
- [x] general acceptance
- [x] make declarations for `data-template="templates:parse-params"` obsolete
- [x] implement #20
- [x] implement #23 as it is a breaking change
- [x] forward port #24

**NOTE:** #22 is out of scope as it can be implement later as a configuration option

Embrace the top-down approach by allowing dynamic includes

with main page template, which is passed to `templates:render`

```html
<html>
    <head>
        <title>${title}</title>
        <script data-template="templates:include"
            data-template-path="${includes?scripts}"></script>
        <style data-template="templates:include"
            data-template-path="${includes?styles}"></style>
    </head>
    <body>
        <!-- include template from module -->
        <nav data-template="templates:include"
            data-template-path="${includes?menubar}"></nav>
        <main data-template="templates:include"
            data-template-path="${includes?main}"></main>
    </body>
</html>
```

and a model, (the second parameter to `templates:render`)

```xquery
map {
  "title": "My Page",
  "data": ((: this can be data loaded from the db for example :)),
  "includes": map {
    "scripts": "templates/includes/scripts.html",
    "styles": "templates/includes/styles.html",
    "main": "templates/pages/my-page.html",
    "menubar": "templates/includes/menubar.html"
  },
  "scripts": ( (: list of scripts to include in the page :) ),
  "styles": ( (: list of stylesheets to include in the page :) )
}
```

It is feasible to create quite flexible template structures that are easier to reason about. You do not need many, maybe no custom template function at all.

Don't worry `templates:surround` is still... around.

## Unified template attributes

An app using this library can now enable this new feature by adding the below key to their configuration:

```xquery
map {
  $templates:CONFIG_ATTR_PREFIX: 'my-prefix'
}
```

Now all template attributes are prefixed with the configured prefix. "my-prefix" in this example.

- `my-prefix-fn` : for templating functions (formerly `data-template`)
- `my-prefix-arg-*` for all templating function arguments (formerly `data-template-*`)
- `my-prefix-target` for block expansion targets AKA slots (formerly `data-target`)

**BREAKING CHANGES**

1. The first parameter to templates:apply has to be a single node now. It was allowed to pass in more than one node but this was untested and the output would have been invalid XML anyway. So this "feature" was dropped.
2. The start and end delimiters for `templates:parse-params` (could be renamed to `templates:substitute`) must be set as configuration options now. This also means there is no way to change them somewhere in between.
3. The option to filter out template attributes from the output is now **on by default**.
    In order to include the in the output you will have to set `$templates:CONFIG_FILTER_ATTRIBUTES : false()` in the options.
4. The lib module was removed. Most of its functionality is now merged into the functions in the templates module. There is one exception: `lib:resolve-apps` is now part of a new module called `tmpl-apps` and the function was renamed to `tmpl-apps:by-abbrev`. It will put all resolved application paths into a map in the model under the `apps` key. Retrieval is done by `<p id="test1">${apps?templating-test}</p>`.

**FEATURES**

Function resolution

You can pass in `xs:QName(?)` as CONFIG_QNAME_RESOLVER instead of passing in a function that calls `fn:function-lookup` CONFIG_FN_RESOLVER. Due to a an issue in eXist-db's XQuery implementation (see https://github.com/eXist-db/exist/issues/4614 ) this can only resolve functions from imported modules at the time of writing. The old approach can be used as before.
If you pass in none of the above the default resolver will still allow you to use all basic template functions in `templates` namespace.

Parameter resolution

The configuration option for parameter resolution can now contain a sequence of functions `(function(xs:string) as item()*)*` where it had to be a single function before. This allows for greater control over where parameter values are read from.
Defaults to 
```xquery
(
    request:get-parameter(?, ()),
    session:get-attribute#1,
    request:get-attribute#1
) 
```

Error reporting

when a template function cannot be resolved then the node is serialised in the error message to make it easier to find where the issue has occurred 

New entry point

`templates:render` is a shorthand to `templates:apply` where the lookup is passed as part of the config

```xquery
templates:render(
    <html><body></body></html>,
    map{},
    map {
        $templates:CONFIG_QNAME_RESOLVER : xs:QName(?),
        $templates:CONFIG_PARAM_RESOLVER : request:get-parameter(?,()),
        $templates:CONFIG_STOP_ON_ERROR : true(),
        $templates:CONFIG_FILTER_ATTRIBUTES : true()
    })

```

Max arity

With `$templates:CONFIG_MAX_ARITY` you can now specify the maximum arity of template functions. Defaults to 8 _was_ 20.

Placeholder replacement

Before you had to declare `data-template="templates:parse-params"` (that was `data-template="lib:parse-params"` respectively in 1.1.0) to have placeholders in attributes and text nodes replaced.
This is no longer necessary and placeholders will always be replaced using the provided resolvers or the model.

**Changes**

- value retrieval => new public function `templates:resolve-key($model, $key)` will call the configured parameter resolvers internally
- lots of refactoring for readability, maintainability and performance

As the tests show this version does raise the minimum required processor version to eXist-db v5.4.1 because of NPEs in v5.2.0.

line-o committed:
next version

line-o committed:
ci: add npm scripts for testing

line-o committed:
several optimizations

BREAKING CHANGES

$templates:lookup-param-from-restserver is now a sequence of functions of arity 1

line-o committed:
errors, ensure configuration is preserved

Error reporting:
When a template function is not defined the error message will include
the serialized node to help users figure out where the problem occurs.

Ensure configuration is present in $model:
Before, each time the configuration was retrieved it was first checked
if it was present.
This check is now moved to the places where users are able to modify
the model and thus may remove it from the map.

- add a new private function templates:process-children
- templates:process will now check for the configuration to be present
- now that the presence of configuration is ensured
  `templates:get-configuration` is no longer necessary as this can be
  expressed with a simple map lookup

line-o committed:
test: split up test cases, re-use document

duncdrum commented:
@line-o please rebase

line-o commented:
@duncdrum done! Thanks for merging the other PRs

line-o committed:
feat: error when included template not found

line-o committed:
chore: remove unused namespace

line-o committed:
chore: add $templates:NS, simplify other QNames

line-o committed:
refactor: templates:form-control for readability

Each case is now a private function of its own (form, input, select).

line-o committed:
refactor: call get-app-root only if necessary

line-o committed:
chore: remove unused, private function

line-o committed:
refactor: templates:surround-options

line-o committed:
chore: add return types, fix formatting

line-o commented on content/templates.xqm:
```diff
@@ -43,13 +45,17 @@ declare variable $templates:ATTR_DATA_TEMPLATE := "data-template";
-declare function templates:apply($content as node()+, $resolver as function(xs:string, xs:int) as item()?, $model as map(*)?) {
-    templates:apply($content, $resolver, $model, ())
+declare function templates:apply(
+    $template as node(),
```
Allowing for only one (main) template to be rendered at the start allowed for some aggressive optimisations. This is also a breaking change

line-o replied:
Though I suspect __no one__ has ever used templates:apply to render multiple nodes at the same time.

joewiz commented:
@line-o p.s. As long as this involves breaking changes, I'm not sure if you want to resurrect a previously reverted breaking change from @wolfgangmm involving the `lib:parse-params()` function? See the details here: https://github.com/eXist-db/templating/issues/8.

line-o commented:
You are right @joewiz !
We should definitely include those in the next major version.

line-o commented:
Also food for discussion is where to put which functions. As it is now possible to use only functions in the templates namespace, which functionality does it have to provide to be useful on its own?

1. [x] rendering a template (`templates:render`, `templates:apply`)
2. [x] loops (`templates:for-each`)
2. [x] subtemplates / partials / blocks (`templates:include`)
    - this was moved to lib but I would like to reverse that or make lib dependency of templates
    - include path should be possible from values in the model
    `<a data-template="templates:include" data-template-path="model(page-template)" />`
4. [x] output of values
    - some directive to allow lookup in the model
    `<a data-template="templates:text" data-template-text="model(link-text)" />`
5. [x] lookup values in the model (see above)
    - some way to tell the parameter resolution that this is a reference to a key in the model and has to be replaced with its value

line-o committed:
feat: allow lookup function again

- refactor initial functions
  - templates:apply#3 and #4
  - templates:render#3
- add function templates:configure called in all initial functions
- attribute filtering function is now selected by configuration instead
  main benefit is the if condition does not need to be called each time

line-o committed:
feat: templates includes all common functions

- lib:include moved to templates:include
- lib:parse-params is now moved to templates:parse-params
- remove lib
- last remaining lib function lib:resolve-apps is now
  apps:by-abbrev
  links are now added to the model under key 'app'
  declare default value for links of apps that were not found
- delimiters are now part of the configuration and cannot be
  changed inside of templates
- add example (and test) for reading template-path from model
- refactor for speed and readability

line-o committed:
chore: ensure compatibility with exist7

- the default handling of duplicates when merging maps will change to
  use-first. Therefore set duplicate
  handling explicitly to use-last where it is crucial

line-o committed:
raise minimum processor version to 5.4.0

This is the first version that implements map:merge#2

line-o committed:
[refactor] restructure functions, fix formatting

line-o committed:
[ignore] fix comment blocks

line-o committed:
[refactor] consolidate variable names

BREAKING CHANGES:
- $templates:START_DELIMITER => $templates:CONFIG_START_DELIMITER
- $templates:END_DELIMITER => $templates:CONFIG_END_DELIMITER

- All errors are prefixed with E_
- All settings are prefixed with CONFIG_

line-o committed:
chore: add array import

In an effort to declare all dependencies explicitly.

line-o committed:
chore: declare return type for entry functions

line-o committed:
refactor: templates:get-root

line-o committed:
refactor: templates:surround

- add comment if surround template was not found
- avoid *-selector in XPath expression

line-o committed:
refactor: templates:parse-options

Simplify check for problematic options.

line-o committed:
refactor: templates:select

The $selected variable either contains the attribute setting an
option as selected or an empty sequence.

line-o committed:
refactor: make param resolvers public

This is in order to drop them from being the default in the next step.

line-o committed:
refactor: extract and rename modules

- apps is now tmpl-apps to make it obvious it is not an application module
- tmpl-util contains utility functions that clobber up the core library
  (unused for the moment)

line-o committed:
refactor: use utility module, build regexes

- build regular expressions for parse-params at configuration and re-use
  those throughout the code base
- remove functions from template that are now in tmpl-util
- use tmpl-util:* replacements in templates functions

line-o committed:
feat: tmpl-util:cast differentiates number types

line-o commented:
@joewiz There is no need to reincorporate the reverted changes from #8 as the delimiters can no longer be set in templates but are now set once in the configuration.

https://github.com/eXist-db/templating/blob/a544af7c8c42190cc233a8d879406e89532dda06/test/xqs/test-suite.xql#L305-L306
            
If, however, this is not acceptable then we would have to look into this once more.

yamahito commented:
Hello, sorry to be a bit late to the party!

I had some suggestions which I've previously raised with @joewiz for improving some aspects of the templating system; I've added them here as feature request #20 : is there any mileage in considering a similar solution as part of this next version?

Looking at the code, the implementation still looks fairly straightforward, with the suggested changes to `templates:process` now occurring in `templates:process-element` instead.

duncdrum reviewed (commented):
Looks good to me. We should check the documentation article for templating for necessary adjustments. 

Needs another rebase

wolfgangmm commented:
I wonder if it would make sense to have `parse-params` enabled by default, which means it would be active unless you stop it for some part of the tree, e.g. with `disable-parse`. This would align with other templating engines (e.g. in Nunjucks you use `raw` to stop substitution).

line-o commented:
I appreciate your feedback, honestly. Good stuff here  @duncdrum @wolfgangmm 

- re #20 (@yamahito): It's definitely not too late. Conditionals is one aspect that should be nailed in the standard-library, but I have not fully understood the proposed change and its implications, yet.
- re documentation (@duncdrum): A complete overhaul is likely due. Would you be open to help on that?
- re `parse-params` (@wolfgangmm): I agree that parsing placeholders should be transparent to the user of the library. Can you imagine a case where you would not want placeholders to be replaced? Would you be open for a discussion on how to implement this? Use `expand-blocks` instead of `process-children`?

duncdrum commented:
@line-o i can help with the docs

line-o committed:
refactor: form and input handlers

line-o committed:
chore: refactor templates:if-parameter-unset

line-o committed:
chore: refactor templates:expand-text

line-o committed:
feat: transparently replace placeholders

This deprecates templates:parse-params as placeholders in
text and attribute nodes will always be replaced.

Before:  `<span data-template="templates:parse-params">[[item]]</span>`

Now : `<span>[[item]]</span>`

Any data-template attribute with the value "templates:parse-params" is
ignored.

line-o committed:
chore(test): fix formatting and whitespace

line-o commented:
@wolfgangmm your proposal is now part of this PR

line-o commented:
I have an additional question regarding the handling of the now obsolete declaration of `data-template="lib:parse-params"`.
Should templates that are rendered with this new version of templating that still include it throw or the declaration just be ignored?

- if we throw than the legacy declarations will certainly be removed in templates
- if we ignore those templates might be usable in 1.x.x and 2.x.x versions of templating (just the templates not the view module)

Is it worth it keeping this extra check in the library or even harmful as those declarations stay in there?
I am leaning towards throwing an error rather than ignoring the declaration.

duncdrum commented:
+1 for throwing an error, explicit design >> black box design

line-o committed:
fix: do not ignore legacy template declarations

Always throw an error if a declared template function cannot be found.
The attempt to ignore certain function names
creates more problems than it solves as it does not
incentivise to clean up templates when switching to
the newer templating library.
The test for templates with a legcacy declaration was apapted.

line-o committed:
test: adapt tests to exist 7.0.0-SNAPSHOT

eXist-db will now return status code 500 server error when an exeption
is thrown during query execution (see
https://github.com/eXist-db/exist/issues/4639).

Since the templating library is not concerned with routing
we simply add this other accepted status code to the list.
Tests expecting errors will pass with both satus code 400 and 500.

line-o committed:
ci: improve exist workflow

line-o commented:
@yamahito I decided to make a cut and implemented your original proposal with an additional error-check when both `data-template-use-when` _and_ `data-template-use-unless` are set on a single element.

line-o committed:
fix(templates): correct variable in templates:surround

line-o committed:
build: improve test app build

line-o committed:
fix: better support for legacy class syntax

line-o committed:
feat: allow templates:each to iterate over array

line-o committed:
chore: remove unused node dependency yeoman-assert

line-o committed:
chore: declar standards compliant output options for html5

line-o committed:
feat: all template attributes are prefixed when configured

```
map {
  $templates:CONFIG_ATTR_PREFIX: 'my-prefix'
}
```

allows to switch to a newer version of templating where all template attributes are prefixed with the configured prefix. "my-prefix" in this example.

- `my-prefix-fn` : for templating functions (formerly `data-template`)
- `my-prefix-arg-*` for all templating function arguments (formerly `data-template-*`)
- `my-prefix-target` for block expansion targets AKA slots (formerly `data-target`)

line-o committed:
test: add dynamic include to test/app/next

line-o committed:
doc: improve xqdoc for two private functions

line-o committed:
chore: refactor and cleanup templates and tmpl-util

- templates:get-template-function-legacy uses tmpl-util function and $config
- simplify templates:call-by-introspection
- unify 4 functions into templates:parameters-from-attributes
- add tmpl-util:first-qname-like
- remove unused functions in tmpl-util

duncdrum commented:
@line-o should this be released ahead of the beta? What's is the status?

line-o commented:
@duncdrum sorry for the radio silence. We should not merge this until it is fully backwards compatible as there are so many applications using this library.
